api/layer/object.go

```diff
@@ -478,20 +486,35 @@ func (n *layer) listSortedObjectsFromNeoFS(ctx context.Context, p allObjectParam
 	return objects, nil
 }
 
-func (n *layer) getAllObjectsVersions(ctx context.Context, bkt *cache.BucketInfo, prefix, delimiter string) (map[string]*objectVersions, error) {
-	ids, err := n.objectSearch(ctx, &findParams{cid: bkt.CID})
-	if err != nil {
-		return nil, err
+func (n *layer) getAllObjectsVersions(ctx context.Context, bkt *api.BucketInfo, prefix, delimiter string) (map[string]*objectVersions, error) {
+	var err error
+
+	cacheKey := cache.CreateObjectsListCacheKey(bkt.CID, prefix)
+	ids := n.listsCache.Get(cacheKey)
+
+	if ids == nil {
+		ids, err = n.objectSearch(ctx, &findParams{cid: bkt.CID, prefix: prefix})
+		if err != nil {
+			return nil, err
+		}
+		if err := n.listsCache.Put(cacheKey, ids); err != nil {
+			n.log.Error("couldn't cache list of objects", zap.Error(err))
+		}
 	}
 
 	versions := make(map[string]*objectVersions, len(ids)/2)
-	for _, id := range ids {
-		meta, err := n.objectHead(ctx, bkt.CID, id)
-		if err != nil {
-			n.log.Warn("could not fetch object meta", zap.Error(err))
+
+	for i := 0; i < len(ids); i++ {
+		if ids[i] == nil {
 			continue
 		}
-		if oi := objectInfoFromMeta(bkt, meta, prefix, delimiter); oi != nil {
+		obj := n.objectFromObjectsCacheOrNeoFS(ctx, bkt.CID, ids[i])
+		if obj == nil {
+			// mark ids[i] as an address of invalid object
+			ids[i] = nil
```

Comment:
If I understand correctly we don't use this mark. So can we remove this line? Or it's assumed we change list in the cache such way (should we prevent concurrent read/write then)?

Reply:
It changes the list but you're right it causes data race. Changed the way to invalidate cache entries.